source/kernel/src/registry/mod.rs

```diff
@@ -0,0 +1,440 @@
+        RD::Request: Serialize + DeserializeOwned,
+        RD::Response: Serialize + DeserializeOwned,
+    {
+        if self.items.iter().any(|i| i.key == RD::UUID) {
+            return Err(());
+        }
+        self.items
+            .push(RegistryItem {
+                key: RD::UUID,
+                value: RegistryValue {
+                    req_resp_tuple_id: RD::type_id(),
+                    req_prod_leaked: kch.clone().leak_erased(),
+                    req_deser: Some(map_deser::<RD>),
+                },
+            })
+            .map_err(drop)
+    }
+
+    /// Get a kernelspace (including drivers) handle of a given driver service.
+    ///
+    /// This can be used by drivers and tasks to interface with a registered driver
+    /// service.
+    ///
+    /// The driver service MUST have already been registered using [Registry::set] or
+    /// [Registry::set_konly] prior to making this call, otherwise no handle will
+    /// be returned.
+    pub fn get<RD: RegisteredDriver>(&self) -> Option<KernelHandle<RD>> {
+        let item = self.items.iter().find(|i| i.key == RD::UUID)?;
+        if item.value.req_resp_tuple_id != RD::type_id() {
+            return None;
+        }
+        unsafe {
+            Some(KernelHandle {
+                prod: item.value.req_prod_leaked.clone_typed(),
+            })
+        }
+    }
+
+    /// Get a handle capable of processing serialized userspace messages to a
+    /// registered driver service.
+    ///
+    /// The driver service MUST have already been registered using [Registry::set] or
+    /// prior to making this call, otherwise no handle will be returned.
+    ///
+    /// Driver services registered with [Registry::set_konly] cannot be retrieved via
+    /// a call to [Registry::get_userspace].
+    pub fn get_userspace<RD>(&mut self) -> Option<UserspaceHandle>
+    where
+        RD: RegisteredDriver,
+        RD::Request: Serialize + DeserializeOwned,
+        RD::Response: Serialize + DeserializeOwned,
+    {
+        let item = self.items.iter().find(|i| &i.key == &RD::UUID)?;
+        Some(UserspaceHandle {
+            req_producer_leaked: item.value.req_prod_leaked.clone(),
+            req_deser: item.value.req_deser?,
+        })
+    }
+}
+
+// UserRequest
+
+// UserResponse
+
+impl<U: MaxSize> MaxSize for UserResponse<U> {
+    const POSTCARD_MAX_SIZE: usize = {
+        <[u8; 16] as MaxSize>::POSTCARD_MAX_SIZE
+            + <u32 as MaxSize>::POSTCARD_MAX_SIZE
+            + <Result<U, ()> as MaxSize>::POSTCARD_MAX_SIZE
+    };
+}
+
+// Envelope
+
+// Message
+
+// ReplyTo
+
+impl<U> ReplyTo<U> {
+    pub async fn reply_konly(self, payload: Result<U, ()>) -> Result<(), ()> {
+        let hmsg = Envelope { body: payload };
+        match self {
+            ReplyTo::KChannel(kprod) => kprod.enqueue_async(hmsg).await.map_err(drop),
+            ReplyTo::Rosc(sender) => sender.send(hmsg),
+            ReplyTo::Userspace { .. } => Err(()),
```

Comment:
nit, would be nice if error values were not discarded here; the error from a failed enqueue on a `KChannel` is _very_ different from "you tried to `reply_konly` to a `Userspace` message", and i would expect drivers to be able to want to handle those separately?